tool/gravity/cli/clusterupdate.go

```diff
@@ -222,8 +224,86 @@ func executeUpdatePhaseForOperation(env *localenv.LocalEnvironment, environ Loca
 		return trace.Wrap(err)
 	}
 	defer updater.Close()
-	err = updater.RunPhase(context.TODO(), params.PhaseID, params.Timeout, params.Force)
-	return trace.Wrap(err)
+	return executeOrForkPhase(env, updater, params, operation)
+}
+
+// gravityResumeServiceName is the name of systemd service that executes
+// the gravity resume command.
+const gravityResumeServiceName = "gravity-resume.service"
+
+// executeOrForkPhase either directly executes the specified operation phase,
+// or launches a one-shot systemd service that executes it in the background.
+func executeOrForkPhase(env *localenv.LocalEnvironment, updater updater, params PhaseParams, operation ops.SiteOperation) error {
+	// If given the --block flag, we're running as a systemd unit (or a user
+	// requested the command to execute in foreground), so proceed to perform
+	// the command (resume or single phase) directly.
+	if params.Block {
+		return updater.RunPhase(context.TODO(),
+			params.PhaseID,
+			params.Timeout,
+			params.Force)
+	}
+	// Before launching the service, perform a few prechecks, for example to
+	// make sure that the operation is being resumed from the correct node.
+	//
+	// TODO(r0mant): Also, make sure agents are running on the cluster nodes:
+	// https://github.com/gravitational/gravity/issues/1667
+	if err := updater.Check(params.toFSM()); err != nil {
+		return trace.Wrap(err)
+	}
+	// Make sure to launch the unit command with the --block flag.
+	args := append(os.Args[1:], "--debug", "--block")
+	env.PrintStep("Starting %v service", gravityResumeServiceName)
+	if err := launchOneshotService(gravityResumeServiceName, args); err != nil {
+		return trace.Wrap(err)
+	}
+	env.PrintStep(`Service %[1]v has been launched.
+
+To monitor the operation progress:
+
+  sudo gravity plan --operation-id=%[2]v --tail
+
+To monitor the service logs:
+
+  sudo journalctl -u %[1]v -f
+`, gravityResumeServiceName, operation.ID)
+	return nil
+}
+
+// launchOneshotService launches the specified command as a one-shot systemd
+// service with the specified name.
+func launchOneshotService(name string, args []string) error {
+	systemd, err := systemservice.New()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// See if the service is already running.
+	status, err := systemd.StatusService(name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Since we're using a one-shot service, it will be in "activating" state
+	// for the duration of the command execution. In fact, one-shot services
+	// never reach "active" state, but we're checking it too just in case.
+	switch status {
+	case systemservice.ServiceStatusActivating, systemservice.ServiceStatusActive:
+		return trace.AlreadyExists("service %v is already running", name)
```

Comment:
Hmm, active means the service has already run?
Activating also does not mean the same service - could be one from previous attempts with a different command line.

Reply:
> active means the service has already run

One-shot services never reach the "active" state. They go straight from "activating" to "exited" (or whatever the terminal state is called).

> Activating also does not mean the same service

Which is fine, since we don't want to launch another resume while another one is going, even if it's a different one.

Reply:
To clarify a bit more on the one-shot services and "active" state - they never reach it unless `RemainAfterExit=true` is set (which it is not the for the "gravity-resume" service). From systemd docs on one-shot services:

> Note that if this option is used without RemainAfterExit= the service will never enter "active" unit state, but directly transition from "activating" to "deactivating" or "dead" since no process is configured that shall run continuously. In particular this means that after a service of this type ran (and which has RemainAfterExit= not set) it will not show up as started afterwards, but as dead.

For "gravity-agent" service we do set `RemainAfterExit=true` for some reason - I don't know what was the reasoning behind it, so the agent service indeed shows as "active" after successful upgrade, which is weird.